### PR TITLE
[PAC][AArch64] Lower ptrauth constants in data

### LINKED ISSUE #83789: [PAC][ELF] Implement codegen support for signed constants
After https://github.com/llvm/llvm-project/issues/83788 the downstream implementation that uses wrapper globals should be rewritten to use special `ptrauth` constant type

### PR DESCRIPTION
**Pull request by kovdan01**
Lower global references to ptrauth constants into `@AUTH` `MCExpr`'s.
The logic is common for MachO and ELF - test both.

Co-authored-by: Ahmed Bougacha <ahmed@bougacha.org>

### CODE REVIEW

**Review by ahmedbougacha — APPROVED** [review]

File: llvm/include/llvm/CodeGen/AsmPrinter.h [hunk 1]
```diff
@@ -585,6 +586,11 @@ class AsmPrinter : public MachineFunctionPass {
     emitGlobalConstant(DL, CV);
   }
 
+  /// Lower the specified ptrauth constant to an MCExpr.
```
Comment (fmayer) [inline llvm/include/llvm/CodeGen/AsmPrinter.h:589]: nit: This comment doesn't add anything over the function prototype

File: llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp [hunk 2]
```diff
@@ -1575,6 +1579,50 @@ void AArch64AsmPrinter::emitPtrauthBranch(const MachineInstr *MI) {
...
+  uint64_t KeyID = CPA.getKey()->getZExtValue();
+  // We later rely on valid KeyID value in AArch64PACKeyIDToString call from
+  // AArch64AuthMCExpr::printImpl, so fail fast.
+  if (KeyID > AArch64PACKey::LAST)
+    report_fatal_error("invalid AArch64 PAC Key ID '" + Twine(KeyID) + "'");
+
+  uint64_t Disc = CPA.getDiscriminator()->getZExtValue();
+  if (!isUInt<16>(Disc))
+    report_fatal_error("invalid AArch64 PAC Discriminator '" + Twine(Disc) +
+                       "'");
```
Comment (fmayer) [inline llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:1619]: nit: "expected uint16"?

File: llvm/include/llvm/CodeGen/AsmPrinter.h (hunk 1 above)
Comment (kovdan01) [reply llvm/include/llvm/CodeGen/AsmPrinter.h]: Makes sense, deleted the comment, see 9633e04f21f0745dc76f905684168333c2c4ff4d

File: llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp (hunk 2 above)
Comment (kovdan01) [reply llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp]: Thanks, I've changed the error message here and above to "AArch64 PAC Discriminator 'XXX' out of range [0, 0xFFFF]" - this looks like a common format for similar error messages across llvm. See 9633e04f21f0745dc76f905684168333c2c4ff4d

**Comment by kovdan01** [comment]
@fmayer Thanks for feedback! Do I get it right that you want me **not** to merge the PR until @pcc looks through the changes? Alternatively, I can merge this "as is" since we already have an approve, and, if there are some issues, just submit a follow-up patch later. This one is a prerequisite for PR94241, so it would be nice to get this merged relatively soon.

**Comment by kovdan01** [comment]
@fmayer @pcc Please let me know if you want to leave some feedback before this is merged or if I can merge it "as is"

**Comment by kovdan01** [comment]
Pushed merge commit 20ed2249689df31f39f9f0a46e0f8269ed0a1039 to re-trigger builkite (previously failed due to unrelated reasons)

**Pull request merged**